File: interview/quick_sort.py
"""题目描述
用python写一个快排并给出时间复杂度
"""

def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    base = arr[len(arr) // 2]
    left = [x for x in arr if x < base]
    middle = [x for x in arr if x == base]
    right = [x for x in arr if x > base]
    return quick_sort(left) + middle + quick_sort(right)

arr = [3, 6, 8, 10, 1, 2, 1]
print(quick_sort(arr))

File: src/interview/quickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
    public:
    vector<int> sortArray(vector<int>&nums){
        if(nums.size() <= 1){
            return nums;
        }
        int base = nums[0];
        vector<int> left, middle, right;
        for (int i = 0; i < int(nums.size()); ++i)
        {
            if(nums[i] < base){
                left.push_back(nums[i]);
            }else if(nums[i] == base){
                middle.push_back(nums[i]);
            }else{
                right.push_back(nums[i]);
            }
        }
        left = sortArray(left);
        right = sortArray(right);
        left.insert(left.end(), middle.begin(), middle.end());
        left.insert(left.end(), right.begin(), right.end());
        return left;
    }
};

int main()
{
    Solution su;
    vector<int> nums = {5, 2, 3, 1};
    vector<int> sorted = su.sortArray(nums);
    for (int num : sorted) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

